mlrun/package/context_handler.py

```diff
@@ -0,0 +1,355 @@
+        # Index 2-...: If it is from mlrun.runtimes we can be sure it ran via MLRun, otherwise not.
+        if self._context is None:
+            for callstack_frame in inspect.getouterframes(inspect.currentframe()):
+                if (
+                    os.path.join("mlrun", "runtimes", "local")
+                    in callstack_frame.filename
+                ):
+                    import mlrun
+
+                    self._context = mlrun.get_or_create_ctx("context")
+                    break
+
+        # If a context was found, give the packagers manager custom packagers to collect (if available):
+        if self._context is not None:
+            # TODO: Complete this in the project spec first
+            # self._packagers_manager.collect_packagers(packagers=self._context.project.get_custom_packagers())
+            pass
+
+    def is_context_available(self) -> bool:
+        """
+        Check if a context was found by the method `look_for_context`.
+
+        :returns: True if a context was found and False otherwise.
+        """
+        return self._context is not None
+
+    def parse_inputs(
+        self,
+        args: tuple,
+        kwargs: dict,
+        type_hints: OrderedDict,
+    ) -> tuple:
+        """
+        Parse the given arguments and keyword arguments data items to the expected types.
+
+        :param args:       The arguments tuple passed to the function.
+        :param kwargs:     The keyword arguments dictionary passed to the function.
+        :param type_hints: An ordered dictionary of the expected types of arguments.
+
+        :returns: The parsed args (kwargs are parsed inplace).
+        """
+        # Parse the type hints (in case some were given as strings):
+        type_hints = {
+            key: self.parse_type_hint(type_hint=value)
+            for key, value in type_hints.items()
+        }
+
+        # Parse the arguments:
+        parsed_args = []
+        type_hints_keys = list(type_hints.keys())
+        for i, argument in enumerate(args):
+            if isinstance(argument, DataItem) and type_hints[
+                type_hints_keys[i]
+            ] not in [
+                inspect._empty,
+                DataItem,
+            ]:
+                parsed_args.append(
+                    self._packagers_manager.unpack(
+                        data_item=argument,
+                        type_hint=type_hints[type_hints_keys[i]],
+                    )
+                )
+                continue
+            parsed_args.append(argument)
+        parsed_args = tuple(parsed_args)  # `args` is expected to be a tuple.
+
+        # Parse the keyword arguments:
+        for key in kwargs.keys():
+            if isinstance(kwargs[key], DataItem) and type_hints[key] not in [
+                inspect._empty,
+                DataItem,
+            ]:
+                kwargs[key] = self._packagers_manager.unpack(
+                    data_item=kwargs[key], type_hint=type_hints[key]
```

Comment:
I take ALL function parameters (both args and kwargs) hints (and if there is no hint then a default value is replacing it: `inspect._empty`) so the length of the parameters taken is always >= then the type hints available or given by the user.

```python
# Get the available parameters type hints from the function's signature:
                func_signature = inspect.signature(func)
                parameters = OrderedDict(
                    {
                        parameter.name: parameter.annotation
                        for parameter in func_signature.parameters.values()
                    }
                )
                # If user input is given, add it on top of the collected defaults (from signature):
                if isinstance(inputs, dict):
                    parameters.update(inputs)
                inputs = parameters
```

And, because I use `OrderedDict` I can use enumeration for args and continue with kwargs.